icu4j/main/core/src/main/java/com/ibm/icu/util/IslamicCalendar.java

```diff
@@ -319,6 +320,294 @@ public IslamicCalendar(int year, int month, int date, int hour,
+                ms+= monthLength(year, i);
+            }
+            return ms;
+        }
+        public int monthLength(int extendedYear, int month) {
+            if (extendedYear < UMALQURA_YEAR_START || extendedYear > UMALQURA_YEAR_END) {
+                return CIVIL_ALGORITHM.monthLength(extendedYear, month);
+            }
+            int index = (extendedYear - UMALQURA_YEAR_START);     // calculate year offset into bit map array
+            int mask = (0x01 << (11 - month));                  // set mask for bit corresponding to month
+            if((UMALQURA_MONTHLENGTH[index] & mask) != 0) {
+                return 30;
+            }
+            return 29;
+        }
+        public int yearLength(int extendedYear) {
+            if (extendedYear < UMALQURA_YEAR_START  || extendedYear > UMALQURA_YEAR_END) {
+                return CIVIL_ALGORITHM.yearLength(extendedYear);
+            }
+            int length = 0;
+            for(int i = 0; i < 12; i++) {
+                length += monthLength(extendedYear, i);
+            }
+            return length;
+        }
+        public int[] compute(long julianDays, long current) {
+            long days = julianDays - epoch();
+            if( days < yearStart(UMALQURA_YEAR_START)) {
+                return CIVIL_ALGORITHM.compute(julianDays, current);
+            }
+            // Estimate a value y which is closer to but not greater than the year.
+            // It is the inverse function of the logic inside yearStart() about the
+            // linear estimate.
+            int year = (int)((days - (460322.05 + 0.5)) / 354.36720) + UMALQURA_YEAR_START - 1;
+            int month = 0;
+            long monthStart;
+            long d = 1;
+            while (d > 0) {
+                year++;
+                d = days - yearStart(year) +1;
+                int yearLength = yearLength(year);
+                if (d == yearLength) {
+                    month = 11;
+                    break;
+                } else if (d < yearLength) {
+                    int monthLen = monthLength(year, month);
+                    for (month = 0; d > monthLen; monthLen = monthLength(year, ++month)) {
+                        d -= monthLen;
+                    }
+                    break;
+                }
+            }
+            int[] values = {
+              year,
+              month,
+              (int)(days - monthStart(year, month)) + 1,
+              (int)(days - yearStart(year) + 1)};
+            return values;
+        }
+    };
+
+    private static Algorithm ISLAMIC_ALGORITHM;
+    private static Algorithm CIVIL_ALGORITHM;
+    private static Algorithm TBLA_ALGORITHM;
+    private static Algorithm UMALQURA_ALGORITHM;
+
+    static {
+        ISLAMIC_ALGORITHM = new IslamicAlgorithm();
+        CIVIL_ALGORITHM = new CivilAlgorithm();
+        TBLA_ALGORITHM = new TBLAAlgorithm();
+        UMALQURA_ALGORITHM = new UmalquraAlgorithm();
+    };
```

Comment:
Just putting these initializers on the member definitions above doesn't work?

Reply:
no, you cannot do that in java
https://www.baeldung.com/java-static-instance-initializer-blocks